fix(routing): Improve project switching for insights pages

HamedMP opened the pull request:
## Problem

Right now, when switching a project while being inside an insight, it just updates the project id:
e.g: from 
`https://us.posthog.com/project/80848/insights/rwQGkcYU` to
`https://us.posthog.com/project/19618/insights/rwQGkcYU`

but since the insight is not there, it shows "Insight not found".
![image](https://github.com/user-attachments/assets/cda6556b-38b2-4430-a1a6-660ae0446532)


<!-- Who are we building for, what are their needs, why is this important? -->

General UX improvement

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

When changing the project from the project switcher, check if it's in the insight page and in that case, route to the `[new project id]/insights` page instead of `[new project id]/insights/ [old insight id/`

This changes the #25009 PR by @Twixes. 

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Does this work well for both Cloud and self-hosted?
Yes
<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

Locally by creating a new project and switching between them while being inside an insight. 
<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

[Slack thread](https://posthog.slack.com/archives/C0368RPHLQH/p1739522928079419)

HamedMP committed:
fix: Improve project switching for insights pages

Twixes reviewed (changes requested):
Thanks! So there's a couple things to make this work reliably:
- Make sure we're only redirecting away from the resource if switching to an env of a _different project_ (easy to check with `projectLogic`)
- Build the smart redirect in a more abstract way, since this problem applies to every single resource, not just insights. A good way could be taking the URL of the parent breadcrumb of the current resource's scene (`breadcrumbsLogic`). We'll just have to maintain a mapping of which scenes represent a project-level (or environment-level) resource

HamedMP committed:
fix: Improve project switching with breadcrumb-aware navigation

HamedMP commented:
@Twixes your suggestions are implemented in the last commit. 

Now:
1. switching between different envs of the same team keeps the full url, only changing the projectId.
2. switching between different teams goes to the root of the resource.

Twixes commented on frontend/src/layout/navigation/ProjectSwitcher.tsx:
```diff
@@ -112,8 +110,19 @@ function OtherProjectButton({ team }: { team: TeamBasicType; onClickInside?: ()
 
+        // If switching between projects (not just environments), redirect to parent path
+        if (currentTeam && currentTeam.project_id !== team.project_id) {
+            // Get the parent breadcrumb's path if it exists
+            const parentBreadcrumb = breadcrumbs.length > 1 ? breadcrumbs[breadcrumbs.length - 2] : null
+            if (parentBreadcrumb?.path) {
+                return urls.project(team.id, parentBreadcrumb.path)
+            }
+            // If no parent path found, go to project home
+            return urls.project(team.id)
```
This doesn't quite work in the expected way, because it ALWAYS takes the URL a level up, even if already in a non-resource-specific scene. E.g. switching projects when viewing http://localhost:8010/project/2/insights result in the user seeing the project _homepage_ after switching (http://localhost:8010/project/2)

HamedMP replied:
Yes, that's right, i can see pros and cons for both approach.

Do you think it helps to keep the `/insights` path when switching projects? It might be that going to home (and seeing the default Dashboard) is more expected than being put in a sub-page of the new project. 
(since it's completely a new project (let's say `Project B`, most likely i'm switching to explore something quite different than the current resource i'm viewing in `Project A`).

Twixes replied:
In my personal experience, I would expect to still see insights, especially if I went there and only realized I'm in the wrong project when seeing the list.

HamedMP replied:
Yes, makes sense. I also wonder if we can do a feature flag for this, and see how many times people go to another page after switching projects (or maybe we can already check this from our current events), wdyt?

Twixes reviewed (changes requested):
Great direction, but commented on a regression from the current experience!

HamedMP commented:
Adding [this](https://posthog.slack.com/archives/C08BU1FEAGP/p1739791564532589) as context here

HamedMP committed:
- **New Utility Function**: Added `getProjectSwitchTargetUrl` in `router-utils.ts` to centralize URL transformation logic
- **Improved Same-Project Navigation**: When switching between teams in the same project, resource IDs are now preserved
- **Configurable Path Handling**:
  - Paths in `STAY_ON_SAME_PAGE_PATHS` (e.g., `settings`) maintain their full path structure
  - Paths in `REDIRECT_TO_PROJECT_ROOT_PATHS` (e.g., `products`, `onboarding`) redirect to project root
  - Other paths with subresources:
    - Within same project: Keep full path including resource ID
    - Between different projects: Go to parent resource

1. **Same Project, Different Environment**:
   - From: `/project/1/insights/abc123` (Project A, Env 1)
   - To: `/project/2/insights/abc123` (Project A, Env 2)

2. **Different Projects**:
   - From: `/project/1/insights/abc123` (Project A)
   - To: `/project/2/insights` (Project B)

3. **Settings Pages** (always preserve full path):
   - From: `/project/1/settings/project`
   - To: `/project/2/settings/project`

4. **Special Routes** (redirect to project root):
   - From: `/project/1/products`
   - To: `/project/2`

- Removed duplicate URL handling logic from `EnvironmentSwitcher.tsx` and `ProjectSwitcher.tsx`
- Added project ID comparison to make informed decisions about preserving resource IDs
- Improved type safety with optional project ID parameters

HamedMP commented:
## Latest Changes

### URL Handling During Project/Environment Switching

- **New Utility Function**: Added `getProjectSwitchTargetUrl` in `router-utils.ts` to centralize URL transformation logic
- **Improved Same-Project Navigation**: When switching between teams in the same project, resource IDs are now preserved
- **Configurable Path Handling**:
  - Paths in `STAY_ON_SAME_PAGE_PATHS` (e.g., `settings`) maintain their full path structure
  - Paths in `REDIRECT_TO_PROJECT_ROOT_PATHS` (e.g., `products`, `onboarding`) redirect to project root
  - Other paths with subresources:
    - Within same project: Keep full path including resource ID
    - Between different projects: Go to parent resource

### Examples

1. **Same Project, Different Environment**:
   - From: `/project/1/insights/abc123` (Project A, Env 1)
   - To: `/project/2/insights/abc123` (Project A, Env 2)

2. **Different Projects**:
   - From: `/project/1/insights/abc123` (Project A)
   - To: `/project/2/insights` (Project B)

3. **Settings Pages** (always preserve full path):
   - From: `/project/1/settings/project`
   - To: `/project/2/settings/project`

4. **Special Routes** (redirect to project root):
   - From: `/project/1/products`
   - To: `/project/2`

### Code Improvements

- Removed duplicate URL handling logic from `EnvironmentSwitcher.tsx` and `ProjectSwitcher.tsx`
- Added project ID comparison to make informed decisions about preserving resource IDs
- Improved type safety with optional project ID parameters

HamedMP committed:
refactor: Simplify team project ID lookup in environment switcher

- Replace complex iteration over sorted projects map with direct team lookup
- Use `currentOrganization.teams` for more direct and efficient team finding
- Remove unnecessary nested loops and early break condition
- Improve readability and performance of project team identification logic

Twixes reviewed (approved):
Nice

HamedMP committed:
Merge branch 'master' into hamed-fix-project-switcher-redirect-to-insights
